teachertool/src/components/EvalResultDisplay.tsx

```diff
@@ -1,27 +1,46 @@
 /// <reference path="../../../built/pxtblocks.d.ts"/>
 
-import { useContext, useState } from "react";
+import { useContext } from "react";
 import { AppStateContext } from "../state/appStateContext";
+import { getCatalogCriteriaWithId } from "../state/helpers";
+import { CriteriaEvaluationResult } from "../types/criteria";
 
 interface IProps {}
 
 export const EvalResultDisplay: React.FC<IProps> = ({}) => {
     const { state: teacherTool } = useContext(AppStateContext);
 
+    function getTemplateStringFromCriteriaInstanceId(instanceId: string): string {
+        const catalogCriteriaId = teacherTool.selectedCriteria?.find(
+            criteria => criteria.instanceId === instanceId
+        )?.catalogCriteriaId;
+        if (!catalogCriteriaId) return "";
+        return getCatalogCriteriaWithId(catalogCriteriaId)?.template ?? "";
+    }
+
     return (
         <>
             {teacherTool.projectMetadata && (
                 <div className="eval-results-container">
                     <h3>{lf("Project: {0}", teacherTool.projectMetadata.name)}</h3>
-                    {teacherTool.currentEvalResult === undefined && <div className="common-spinner" />}
-                    {Object.keys(teacherTool.currentEvalResult?.blockIdResults ?? {}).map(id => {
-                        const result = teacherTool.currentEvalResult?.blockIdResults[id];
+
+                    {teacherTool.evalResults.length === 0 && <div className="common-spinner" />}
+                    {Object.keys(teacherTool.evalResults ?? {}).map(criteriaInstanceId => {
+                        const result = teacherTool.evalResults[criteriaInstanceId];
+
                         return (
-                            <div className="result-block-id" key={id}>
-                                <p className="block-id-label">{id}:</p>
-                                <p className={result ? "positive-text" : "negative-text"}>
-                                    {result ? "passed" : "failed"}
+                            <div className="result-block-id" key={criteriaInstanceId}>
+                                <p className="block-id-label">
+                                    {getTemplateStringFromCriteriaInstanceId(criteriaInstanceId)}:
                                 </p>
+                                {result === CriteriaEvaluationResult.InProgress && <div className="common-spinner" />}
```

Comment:
This was another case where this UI is all going to change dramatically once we start working on the results page, so I just did something quick for now.

But to answer your immediate question: since InProgress is just another status for the result, I don't think it could be in both "InProgress" and "Fail" at the same time.